### PR TITLE
fix: js library default export to a unique name instead of default key in self

### LINKED ISSUE #20572: [Bug]: Library namespace randomly changes to a different name and stops working
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When I install a library and start using it, the namespace randomly changes when I head to the homepage and open the app again. 

Discord user report: https://discord.com/channels/725602949748752515/1072952744509448192

### Steps To Reproduce

1. Install jspdf-autotable by using this URL- https://cdn.jsdelivr.net/npm/jspdf-autotable@3.5.28/dist/jspdf.plugin.autotable.min.js
2. The library gets installed with the namespace `default`
3. Head back to the home page
4. Open the app again
5. See that the library's namespace has now changed to `__createTable`

### Public Sample App

_No response_

### Issue video log

Loom recording: https://www.loom.com/share/7cdbcfed7f8e4acb9fb4a81234da80dc

### Version

Cloud

Front logo [Front conversations](https://app.frontapp.com/open/top_7dxos)

### PR DESCRIPTION
**Pull request by AmanAgarwal041**
## Description
Earlier since some of the libraries one of which is `https://cdn.jsdelivr.net/npm/jspdf-autotable@3.5.28/dist/jspdf.plugin.autotable.js` wasn't exporting the functionality to some key, instead it was exporting as default. Hence, after the import of the library the functionality was being set on the default key, which on refresh or on reintializing the application in the store got changed and the inner functions got attached to global object.

This PR fixes this issue and checks if the last accessor is the `default` then it maps all the functionality to a unique name generated from the library url and moves everything to the generated name as done by other libraries.

Fixes #20572
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.JS"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11011513239>
> Commit: b4c076793f4f28b4b84bfd003e916b70f102c773
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11011513239&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.JS`
> Spec:
> <hr>Tue, 24 Sep 2024 10:40:20 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Improved handling of library installations, enhancing the management of library exports and default exports.
	- Added a new test case to verify the installation of libraries with default exports.

- **Bug Fixes**
	- Resolved issues related to key differentiation during library installation.

### CODE REVIEW

File: app/client/src/workers/Evaluation/handlers/jsLibrary.ts [hunk 1]
```diff
@@ -128,10 +128,37 @@
...
       const keysAfterInstallation = Object.keys(self);
 
-      accessors.push(
-        ...difference(keysAfterInstallation, envKeysBeforeInstallation),
+      const differentiatingKeys = difference(
+        keysAfterInstallation,
+        envKeysBeforeInstallation,
       );
 
+      if (differentiatingKeys[differentiatingKeys.length - 1] === "default") {
```
[the remark this answers is not in the document]
Comment (AmanAgarwal041) [reply app/client/src/workers/Evaluation/handlers/jsLibrary.ts]: Addressed!

File: app/client/src/workers/Evaluation/handlers/jsLibrary.ts [hunk 2]
```diff
@@ -128,10 +128,37 @@
...
+        self[uniqueName] = self["default"];
+        // removing default from differentiating keys
+        differentiatingKeys.pop();
+        // deleting the reference of default key from the self object
+        delete self["default"];
+        // mapping all the references of differentiating keys from the self object to the self[uniqueName] key object
+        differentiatingKeys.map((key) => {
+          self[uniqueName][key] = self[key];
+          // deleting the references from the self object
+          delete self[key];
```
[the remark this answers is not in the document]
Comment (AmanAgarwal041) [reply app/client/src/workers/Evaluation/handlers/jsLibrary.ts]: Addressed!

File: app/client/src/workers/Evaluation/handlers/jsLibrary.ts [hunk 3]
```diff
@@ -128,10 +128,37 @@ export async function installLibrary(
...
+          takenAccessors,
+          takenNamesMap,
+        );
+
+        // mapping default functionality to library name accessor
+        self[uniqueName] = self["default"];
+        // removing default from differentiating keys
+        differentiatingKeys.pop();
+        // deleting the reference of default key from the self object
+        delete self["default"];
```
[the remark this answers is not in the document]
Comment (AmanAgarwal041) [reply app/client/src/workers/Evaluation/handlers/jsLibrary.ts]: Addressed!

**Comment by rishabhrathod01** [comment]
@AmanAgarwal041 I tried to reinstall the library after uninstalling it and I was able to reproduce the issue. I demoed the same in the recording below. 
https://github.com/user-attachments/assets/b6722a82-33bd-4cd4-b9f1-e218935e7010

Could you handle this case and add a test for it?

**Review by rishabhrathod01 — CHANGES_REQUESTED** [review]
https://github.com/appsmithorg/appsmith/pull/36483#issuecomment-2370604283

**Comment by AmanAgarwal041** [comment]
@rishabhrathod01 Addressed the issue. It was acting weird because of the `undefined` being set to self property instead of just getting delete and hence on second time there were no differentiating keys in the before and after self.

File: app/client/src/workers/Evaluation/handlers/jsLibrary.ts [hunk 4]
```diff
@@ -128,10 +128,40 @@ export async function installLibrary(
...
+        // mapping all the references of differentiating keys from the self object to the self[uniqueName] key object
+        differentiatingKeys.map((key) => {
+          if (key !== "default") {
+            self[uniqueName][key] = self[key];
+            // deleting the references from the self object
+            delete self[key];
+          }
+        });
```
[the remark this answers is not in the document]
Comment (rishabhrathod01) [reply app/client/src/workers/Evaluation/handlers/jsLibrary.ts:158]: This change caused a bug in the product hence we won't be going ahead with this.

**Review by rishabhrathod01 — APPROVED** [review]
LGTM

**Pull request merged**